lib/broadway.ex

```diff
@@ -587,6 +587,23 @@ defmodule Broadway do
   """
   @type on_start() :: {:ok, pid()} | :ignore | {:error, {:already_started, pid()} | term()}
 
+  @doc """
+  Invoked for preparing messages before handling (if defined).
+
+  It expects:
+
+    * `message` is the `Broadway.Message` struct to be processed.
+    * `context` is the user defined data structure passed to `start_link/2`.
+
+  This is the place to do any pre-processing of the incoming messages, e.g.,
+  filtering/transforming messages according to some criteria.
+
+  This callback is optional. If present, it's called **before** the messages are passed to
+  `c:handle_message/3`. This gives you a change to do something with the messages before
+  they are handled, such as filtering or checking for idempotency.
```

Comment:
We cannot really filter messages, since messages always have to be acknowledged. Here is a suggestion:

```suggestion
  This is the place to prepare and preload any information that will be used
  by `handle_message/2`. For example, if you need to query the database,
  instead of doing it once per message, you can do it on this callback.

  The length of the list of messages received by this callback is based on
  the `min_demand`/`max_demand` configuration in the processor. This callback
  must always return all messages it receives, as `handle_message/2` is still
  called individually for each message afterwards.
```